### DISCUSSION TITLE
Proposal: Allow tuples as attribute parameters

### DISCUSSION THREAD

**Discussion by lukepaw** [dotnet/csharplang#1979] [Language Ideas]
Currently attributes may only use a small set of types as their parameters.
As tuples are deterministic in their layout, as long as they consist of types that can be used as attribute parameters (primitives, strings, arrays of the former), they themselves should also be usable as such (including arrays of tuples).

**Comment by HaloFour** [comment]
This limitation is set by the CLR which would need support for reading the attribute parameters at runtime in order to instantiate the attribute.

**Comment by alrz** [comment]
I'd like this to be relaxed for *any* type constructor,
```cs
[SomeAttribute(Foo(5))]
```
We can already serialize types (for typeof), so I think this shouldn't be impossible to do without clr changes?

**Comment by Joe4evr** [comment]
> We can already serialize types (for typeof [...]

That's done by serializing the type's FQDN in its place and looking it up/injecting it at runtime.

**Comment by ufcpp** [comment]
This could be essentially the same issue as [const ValueTuple](https://github.com/dotnet/csharplang/issues/1034) and [Unmanaged Constants](https://github.com/dotnet/csharplang/issues/688).

**Comment by alrz** [comment]
>That's done by serializing the type's FQDN in its place and looking it up/injecting it at runtime.

That's why I think it's not a clr concern. Only the reflection API needs to be updated to recognize the new format?

**Comment by HaloFour** [comment]
@alrz 

> That's why I think it's not a clr concern. Only the reflection API needs to be updated to recognize the new format?

That's possible.  To me the line between CLR and BCL has always been a little fuzzy when it comes to mscorlib.  Either way it requires a modification outside of the C# language to be supported.
